File: examples/stablecoin-payments/typescript/escrow-payment.ts
/**
 * Escrow-Based Payment Example
 *
 * Demonstrates secure escrow payments for e-commerce transactions:
 * - Create an escrow for a purchase
 * - Hold funds until conditions are met
 * - Release to seller upon delivery confirmation
 * - Handle refunds if needed
 *
 * This pattern is ideal for:
 * - Marketplace transactions
 * - Freelance payments
 * - Any buyer-seller scenario requiring trust
 *
 * Usage: npx ts-node escrow-payment.ts
 */

import { StatesetClient } from '@stateset/core';
import type { EscrowSettlement } from '@stateset/core/types';

// Configuration
const RPC_ENDPOINT = process.env.STATESET_RPC || 'http://localhost:26657';
const BUYER_MNEMONIC = process.env.BUYER_MNEMONIC || 'buyer mnemonic here';
const SELLER_MNEMONIC = process.env.SELLER_MNEMONIC || 'seller mnemonic here';

const SSUSD_DENOM = 'ssusd';
const SSUSD_DECIMALS = 6;

function formatAmount(amount: string): string {
  return (parseInt(amount) / Math.pow(10, SSUSD_DECIMALS)).toFixed(2);
}

function toBaseUnits(amount: number): string {
  return Math.floor(amount * Math.pow(10, SSUSD_DECIMALS)).toString();
}

function printEscrowStatus(escrow: EscrowSettlement) {
  console.log('┌─────────────────────────────────────────────────────────┐');
  console.log('│                    ESCROW STATUS                        │');
  console.log('├─────────────────────────────────────────────────────────┤');
  console.log(`│ ID:         ${escrow.id.padEnd(44)}│`);
  console.log(`│ Status:     ${escrow.status.padEnd(44)}│`);
  console.log(`│ Amount:     ${formatAmount(escrow.amount.amount).padEnd(44)}│`);
  console.log(`│ Depositor:  ${escrow.depositor.slice(0, 20).padEnd(44)}│`);
  console.log(`│ Recipient:  ${escrow.recipient.slice(0, 20).padEnd(44)}│`);
  console.log(`│ Release:    ${escrow.releaseTime.toISOString().slice(0, 19).padEnd(44)}│`);
  console.log('└─────────────────────────────────────────────────────────┘');
}

async function demonstrateEscrowPayment() {
  console.log('═'.repeat(60));
  console.log('  ssUSD Escrow Payment Example');
  console.log('  Secure Marketplace Transaction');
  console.log('═'.repeat(60));
  console.log();

  // Connect both parties
  console.log('1. Connecting buyer and seller...');
  const buyerClient = await StatesetClient.connectWithMnemonic(BUYER_MNEMONIC, {
    rpcEndpoint: RPC_ENDPOINT,
  });
  const sellerClient = await StatesetClient.connectWithMnemonic(SELLER_MNEMONIC, {
    rpcEndpoint: RPC_ENDPOINT,
  });

  const buyerAddress = buyerClient.getAddress();
  const sellerAddress = sellerClient.getAddress();

  console.log(`   Buyer:  ${buyerAddress}`);
  console.log(`   Seller: ${sellerAddress}`);
  console.log();

  // Check initial balances
  console.log('2. Initial Balances:');
  const buyerBalance = await buyerClient.getBalance(SSUSD_DENOM);
  const sellerBalance = await sellerClient.getBalance(SSUSD_DENOM);
  console.log(`   Buyer:  ${formatAmount(buyerBalance)} ssUSD`);
  console.log(`   Seller: ${formatAmount(sellerBalance)} ssUSD`);
  console.log();

  // Create escrow
  const purchaseAmount = 500; // 500 ssUSD for the product
  const escrowDuration = 7 * 24 * 60 * 60 * 1000; // 7 days in ms
  const releaseTime = new Date(Date.now() + escrowDuration);

  console.log('3. Creating Escrow:');
  console.log(`   Purchase Amount: ${purchaseAmount} ssUSD`);
  console.log(`   Release Time:    ${releaseTime.toISOString()}`);
  console.log(`   Conditions:      Delivery confirmation required`);
  console.log();

  console.log('   Submitting escrow transaction...');
  const escrowResult = await buyerClient.settlement.createEscrow({
    recipient: sellerAddress,
    amount: {
      denom: SSUSD_DENOM,
      amount: toBaseUnits(purchaseAmount),
    },
    releaseTime,
    conditions: JSON.stringify({
      type: 'delivery_confirmation',
      orderId: 'ORD-12345',
      requiresTracking: true,
    }),
  });

  console.log(`   ✓ Escrow created!`);
  console.log(`   TX Hash:   ${escrowResult.transactionHash}`);
  console.log(`   Escrow ID: ${escrowResult.escrowId}`);
  console.log();

  // Check escrow status
  console.log('4. Escrow Status:');
  const escrow = await buyerClient.settlement.getEscrow(escrowResult.escrowId);
  if (escrow) {
    printEscrowStatus(escrow);
  }
  console.log();

  // Check balances after escrow
  console.log('5. Balances After Escrow (funds locked):');
  const buyerBalanceAfterEscrow = await buyerClient.getBalance(SSUSD_DENOM);
  const sellerBalanceAfterEscrow = await sellerClient.getBalance(SSUSD_DENOM);
  console.log(`   Buyer:  ${formatAmount(buyerBalanceAfterEscrow)} ssUSD`);
  console.log(`   Seller: ${formatAmount(sellerBalanceAfterEscrow)} ssUSD`);
  console.log(`   Locked: ${purchaseAmount} ssUSD (in escrow)`);
  console.log();

  // Simulate delivery confirmation and release
  console.log('6. Simulating Order Fulfillment...');
  console.log('   → Seller ships product');
  console.log('   → Tracking: TRACK-12345-ABC');
  console.log('   → Buyer confirms delivery');
  console.log();

  // Release escrow
  console.log('7. Releasing Escrow to Seller:');
  const releaseResult = await buyerClient.settlement.releaseEscrow(escrowResult.escrowId);
  console.log(`   ✓ Escrow released!`);
  console.log(`   TX Hash: ${releaseResult.transactionHash}`);
  console.log();

  // Final balances
  console.log('8. Final Balances:');
  const buyerFinalBalance = await buyerClient.getBalance(SSUSD_DENOM);
  const sellerFinalBalance = await sellerClient.getBalance(SSUSD_DENOM);
  console.log(`   Buyer:  ${formatAmount(buyerFinalBalance)} ssUSD`);
  console.log(`   Seller: ${formatAmount(sellerFinalBalance)} ssUSD`);
  console.log();

  // Calculate changes
  const buyerChange = parseInt(buyerFinalBalance) - parseInt(buyerBalance);
  const sellerChange = parseInt(sellerFinalBalance) - parseInt(sellerBalance);
  console.log('   Balance Changes:');
  console.log(`   Buyer:  ${buyerChange >= 0 ? '+' : ''}${formatAmount(buyerChange.toString())} ssUSD`);
  console.log(`   Seller: ${sellerChange >= 0 ? '+' : ''}${formatAmount(sellerChange.toString())} ssUSD`);

  // Cleanup
  buyerClient.disconnect();
  sellerClient.disconnect();

  console.log();
  console.log('═'.repeat(60));
  console.log('  Escrow Payment Completed Successfully!');
  console.log('═'.repeat(60));
}

/**
 * Demonstrate refund flow
 */
async function demonstrateRefundFlow() {
  console.log();
  console.log('═'.repeat(60));
  console.log('  BONUS: Escrow Refund Example');
  console.log('═'.repeat(60));
  console.log();

  const buyerClient = await StatesetClient.connectWithMnemonic(BUYER_MNEMONIC, {
    rpcEndpoint: RPC_ENDPOINT,
  });
  const sellerAddress = 'stateset1seller...';

  // Create escrow
  console.log('Creating escrow for potential refund...');
  const escrowResult = await buyerClient.settlement.createEscrow({
    recipient: sellerAddress,
    amount: {
      denom: SSUSD_DENOM,
      amount: toBaseUnits(200),
    },
    releaseTime: new Date(Date.now() + 3600000),
    conditions: 'Order fulfillment',
  });

  console.log(`Escrow ID: ${escrowResult.escrowId}`);
  console.log();

  // Simulate dispute scenario
  console.log('Scenario: Order cancelled, initiating refund...');

  // Refund escrow (seller must call this, or buyer after release time)
  const refundResult = await buyerClient.settlement.refundEscrow(escrowResult.escrowId);
  console.log(`✓ Escrow refunded!`);
  console.log(`TX Hash: ${refundResult.transactionHash}`);

  buyerClient.disconnect();
}

// Run the demonstration
async function main() {
  await demonstrateEscrowPayment();
  // Uncomment to see refund flow:
  // await demonstrateRefundFlow();
}

main().catch(console.error);

File: examples/stablecoin-payments/python/escrow_payment.py
#!/usr/bin/env python3
"""
Escrow-Based Payment Example

Demonstrates secure escrow payments for e-commerce:
- Create escrow for purchases
- Hold funds until delivery
- Release to seller on confirmation
- Handle refunds

Usage: python escrow_payment.py
"""

import json
import subprocess
import os
import time
from typing import Optional, Dict, Any, List
from dataclasses import dataclass
from decimal import Decimal
from datetime import datetime, timedelta
from enum import Enum


# Configuration
CHAIN_ID = os.getenv("STATESET_CHAIN_ID", "stateset-1")
NODE = os.getenv("STATESET_NODE", "tcp://localhost:26657")
KEYRING = os.getenv("STATESET_KEYRING", "test")
GAS_PRICES = "0.025ustake"
GAS_LIMIT = "500000"
BINARY = os.getenv("STATESET_BINARY", "statesetd")

SSUSD_DENOM = "ssusd"
SSUSD_DECIMALS = 6


class EscrowStatus(Enum):
    ACTIVE = "active"
    RELEASED = "released"
    REFUNDED = "refunded"
    DISPUTED = "disputed"


@dataclass
class Escrow:
    """Escrow details"""
    escrow_id: str
    depositor: str
    recipient: str
    amount: int
    release_time: datetime
    conditions: str
    status: EscrowStatus


@dataclass
class TransactionResult:
    success: bool
    tx_hash: Optional[str] = None
    escrow_id: Optional[str] = None
    error: Optional[str] = None


class EscrowClient:
    """Client for escrow operations"""

    def __init__(self, key_name: str):
        self.key_name = key_name
        self.address = self._get_address()

    def _get_address(self) -> str:
        cmd = [
            BINARY, "keys", "show", self.key_name,
            "--keyring-backend", KEYRING,
            "--output", "json"
        ]
        result = subprocess.run(cmd, capture_output=True, text=True)
        if result.returncode != 0:
            raise Exception(f"Failed to get address: {result.stderr}")
        return json.loads(result.stdout)["address"]

    def _execute_tx(self, *args) -> TransactionResult:
        cmd = [
            BINARY, "tx", *args,
            "--from", self.key_name,
            "--chain-id", CHAIN_ID,
            "--node", NODE,
            "--keyring-backend", KEYRING,
            "--gas", GAS_LIMIT,
            "--gas-prices", GAS_PRICES,
            "--broadcast-mode", "sync",
            "--yes",
            "--output", "json"
        ]
        result = subprocess.run(cmd, capture_output=True, text=True)

        if result.returncode != 0:
            return TransactionResult(success=False, error=result.stderr)

        try:
            data = json.loads(result.stdout)
            return TransactionResult(
                success=True,
                tx_hash=data.get("txhash"),
                escrow_id=data.get("txhash", "")[:16]  # Simplified escrow ID
            )
        except json.JSONDecodeError:
            return TransactionResult(success=False, error="Failed to parse response")

    def _query(self, *args) -> Dict[str, Any]:
        cmd = [
            BINARY, "query", *args,
            "--chain-id", CHAIN_ID,
            "--node", NODE,
            "--output", "json"
        ]
        result = subprocess.run(cmd, capture_output=True, text=True)
        if result.returncode != 0:
            return {}
        return json.loads(result.stdout)

    def get_balance(self, denom: str = SSUSD_DENOM) -> int:
        result = self._query("bank", "balances", self.address)
        for balance in result.get("balances", []):
            if balance["denom"] == denom:
                return int(balance["amount"])
        return 0

    def create_escrow(
        self,
        recipient: str,
        amount: int,
        release_days: int = 7,
        conditions: Optional[Dict] = None
    ) -> TransactionResult:
        """Create an escrow payment"""
        release_time = datetime.now() + timedelta(days=release_days)
        conditions_json = json.dumps(conditions or {})

        escrow_data = json.dumps({
            "recipient": recipient,
            "amount": {"denom": SSUSD_DENOM, "amount": str(amount)},
            "release_time": release_time.isoformat(),
            "conditions": conditions_json
        })

        return self._execute_tx("settlement", "create-escrow", escrow_data)

    def release_escrow(self, escrow_id: str) -> TransactionResult:
        """Release escrow funds to recipient"""
        return self._execute_tx("settlement", "release-escrow", escrow_id)

    def refund_escrow(self, escrow_id: str) -> TransactionResult:
        """Refund escrow funds to depositor"""
        return self._execute_tx("settlement", "refund-escrow", escrow_id)

    def get_escrow(self, escrow_id: str) -> Optional[Escrow]:
        """Get escrow details"""
        result = self._query("settlement", "escrow", escrow_id)
        if not result:
            return None

        escrow_data = result.get("escrow", {})
        return Escrow(
            escrow_id=escrow_data.get("id", escrow_id),
            depositor=escrow_data.get("depositor", ""),
            recipient=escrow_data.get("recipient", ""),
            amount=int(escrow_data.get("amount", {}).get("amount", 0)),
            release_time=datetime.fromisoformat(escrow_data.get("release_time", datetime.now().isoformat())),
            conditions=escrow_data.get("conditions", ""),
            status=EscrowStatus(escrow_data.get("status", "active").lower())
        )


def format_amount(amount: int) -> str:
    return f"{Decimal(amount) / Decimal(10 ** SSUSD_DECIMALS):.2f}"


def to_base_units(amount: float) -> int:
    return int(amount * (10 ** SSUSD_DECIMALS))


def print_box(lines: List[str], title: str = "", width: int = 60):
    """Print a formatted box"""
    print("┌" + "─" * (width - 2) + "┐")
    if title:
        padding = (width - 2 - len(title)) // 2
        print(f"│{' ' * padding}{title}{' ' * (width - 2 - padding - len(title))}│")
        print("├" + "─" * (width - 2) + "┤")
    for line in lines:
        print(f"│ {line:<{width - 4}} │")
    print("└" + "─" * (width - 2) + "┘")


def print_escrow_status(escrow: Escrow):
    """Print formatted escrow status"""
    lines = [
        f"ID:        {escrow.escrow_id}",
        f"Status:    {escrow.status.value}",
        f"Amount:    {format_amount(escrow.amount)} ssUSD",
        f"Depositor: {escrow.depositor[:20]}...",
        f"Recipient: {escrow.recipient[:20]}...",
        f"Release:   {escrow.release_time.strftime('%Y-%m-%d %H:%M')}",
    ]
    print_box(lines, "ESCROW STATUS")


def demo_escrow_payment():
    """Demonstrate escrow payment flow"""
    print("═" * 60)
    print("  ssUSD Escrow Payment Example")
    print("  Secure Marketplace Transaction")
    print("═" * 60)
    print()

    # Setup participants
    print("1. Setting up participants...")
    buyer = EscrowClient("buyer")
    seller = EscrowClient("seller")
    print(f"   Buyer:  {buyer.address}")
    print(f"   Seller: {seller.address}")
    print()

    # Check initial balances
    print("2. Initial Balances:")
    buyer_balance = buyer.get_balance()
    seller_balance = seller.get_balance()
    print(f"   Buyer:  {format_amount(buyer_balance)} ssUSD")
    print(f"   Seller: {format_amount(seller_balance)} ssUSD")
    print()

    # Create escrow
    purchase_amount = 500.0
    print("3. Creating Escrow:")
    print(f"   Purchase Amount: {purchase_amount} ssUSD")
    print(f"   Escrow Duration: 7 days")
    print(f"   Conditions: Delivery confirmation required")
    print()

    print("   Submitting escrow transaction...")
    result = buyer.create_escrow(
        recipient=seller.address,
        amount=to_base_units(purchase_amount),
        release_days=7,
        conditions={
            "type": "delivery_confirmation",
            "order_id": "ORD-12345",
            "requires_tracking": True
        }
    )

    if result.success:
        print(f"   ✓ Escrow created!")
        print(f"   TX Hash:   {result.tx_hash}")
        print(f"   Escrow ID: {result.escrow_id}")
    else:
        print(f"   ✗ Failed: {result.error}")
        return
    print()

    # Check balances after escrow
    print("4. Balances After Escrow (funds locked):")
    buyer_after = buyer.get_balance()
    seller_after = seller.get_balance()
    print(f"   Buyer:  {format_amount(buyer_after)} ssUSD")
    print(f"   Seller: {format_amount(seller_after)} ssUSD")
    print(f"   Locked: {purchase_amount} ssUSD (in escrow)")
    print()

    # Simulate order fulfillment
    print("5. Order Fulfillment:")
    print("   → Seller ships product")
    print("   → Tracking: TRACK-12345-ABC")
    print("   → Buyer confirms delivery")
    print()

    # Release escrow
    print("6. Releasing Escrow to Seller:")
    release_result = buyer.release_escrow(result.escrow_id)

    if release_result.success:
        print(f"   ✓ Escrow released!")
        print(f"   TX Hash: {release_result.tx_hash}")
    else:
        print(f"   ✗ Release failed: {release_result.error}")
    print()

    # Final balances
    print("7. Final Balances:")
    buyer_final = buyer.get_balance()
    seller_final = seller.get_balance()
    print(f"   Buyer:  {format_amount(buyer_final)} ssUSD")
    print(f"   Seller: {format_amount(seller_final)} ssUSD")
    print()

    # Balance changes
    buyer_change = buyer_final - buyer_balance
    seller_change = seller_final - seller_balance
    print("   Balance Changes:")
    print(f"   Buyer:  {'+' if buyer_change >= 0 else ''}{format_amount(buyer_change)} ssUSD")
    print(f"   Seller: {'+' if seller_change >= 0 else ''}{format_amount(seller_change)} ssUSD")
    print()

    print("═" * 60)
    print("  Escrow Payment Completed Successfully!")
    print("═" * 60)


def demo_refund_flow():
    """Demonstrate escrow refund flow"""
    print()
    print("═" * 60)
    print("  BONUS: Escrow Refund Example")
    print("═" * 60)
    print()

    buyer = EscrowClient("buyer")
    seller_address = "stateset1seller..."

    # Create escrow
    print("Creating escrow for potential refund...")
    result = buyer.create_escrow(
        recipient=seller_address,
        amount=to_base_units(200.0),
        release_days=1,
        conditions={"type": "order_fulfillment"}
    )

    if result.success:
        print(f"Escrow ID: {result.escrow_id}")
        print()

        # Scenario: Order cancelled
        print("Scenario: Order cancelled, initiating refund...")
        refund_result = buyer.refund_escrow(result.escrow_id)

        if refund_result.success:
            print(f"✓ Escrow refunded!")
            print(f"TX Hash: {refund_result.tx_hash}")
        else:
            print(f"✗ Refund failed: {refund_result.error}")
    else:
        print(f"Failed to create escrow: {result.error}")


def main():
    demo_escrow_payment()
    # Uncomment to see refund flow:
    # demo_refund_flow()


if __name__ == "__main__":
    main()
